native/pom.xml

```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.mozilla</groupId>
+        <artifactId>jss-parent</artifactId>
+        <version>4.8.0-SNAPSHOT</version>
+    </parent>
+
+    <groupId>org.mozilla</groupId>
+    <artifactId>jss-native</artifactId>
+    <packaging>so</packaging>
```

Comment:
How will the dependency be defined in other projects that use JSS?
Do we need both of these, or just one?
```
<dependency>
    <groupId>org.mozilla</groupId>
    <artifactId>jss</artifactId>
    <version>4.8.0</version>
</dependency>

<dependency>
    <groupId>org.mozilla</groupId>
    <artifactId>jss-native</artifactId>
    <version>4.8.0</version>
    <type>so</type>
</dependency>
```
Or is it possible to have something like this?
```
<dependency>
    <groupId>org.mozilla</groupId>
    <artifactId>jss</artifactId>
    <version>4.8.0</version>
    <type>so</type>
</dependency>
```

Reply:
Don't you just grab:

```
<dependency>
    <groupId>org.mozilla</groupId>
    <artifactId>jss-parent</artifactId>
    <version>4.8.0</version>
</dependency>
```

(from `pom.xml` in the root of the git tree)

That has two modules, and presumably, wouldn't both get pulled in?

Reply:
IIUC we need to specify the `<type>` too. Some examples:
https://www.mojohaus.org/maven-native/native-maven-plugin/usage.html
https://stackoverflow.com/questions/3064045/using-native-dependencies-inside-maven
Since there are two modules here with different types, I'm not sure.

Regardless, should we call the parent module `jss` and the submodules maybe `jss-java` and `jss-native`?

BTW, I'm getting this error (the regular cmake build works fine):
```
$ mvn package
...
Failed tests:   testControlCharacters(org.mozilla.jss.tests.DirStrConverterTest): arrays first differed at element [0]; expected:<20> but was:<12>
  testNullCharacters(org.mozilla.jss.tests.DirStrConverterTest): arrays first differed at element [0]; expected:<20> but was:<12>
  testEmptyString(org.mozilla.jss.tests.DirStrConverterTest): arrays first differed at element [0]; expected:<19> but was:<12>
  testMultibyteCharacters(org.mozilla.jss.tests.DirStrConverterTest): array lengths differed, expected.length=14 actual.length=11; arrays first differed at element [0]; expected:<28> but was:<12>
  testPrintableCharacters(org.mozilla.jss.tests.DirStrConverterTest): arrays first differed at element [0]; expected:<19> but was:<12>

Tests in error: 
  testConstructorDER(org.mozilla.jss.tests.GenericASN1ExtensionTest)
```

Reply:
Maven isn't enabled for testing. I wonder if there's a way to disable it completely?

Tests won't pass under maven because:

 1. This test isn't enabled in the cmake test suite.
 2. Tests generally require java.security to exist for JSS to be loaded.
 3. Tests depend on each other in a way that isn't easy to express in JUnit files/dependencies.
 4. Tests have side-effects.
 5. Not everything in a Java or JUnit tests. There's a lot of custom Java code but also some nss-utils calls as well. 

So it isn't surprising that this is failing.

Reply:
We can skip tests with `-DskipTests`, but now I got a different error during install:
```
$ mvn install -DskipTests
...
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-install-plugin:3.0.0-M1:install (default-install) on project jss-native: NoFileAssignedException: The packaging for this project did not assign a file to the build artifact -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :jss-native
```

Reply:
Found this example (not sure if it will help though):
https://github.com/StephanU/native-maven-plugin-example